### PR TITLE
SALM with NeMo Automodel integration for Nemotron Nano V3 LLM backbone

### PR DESCRIPTION
**Pull request by pzelasko**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add `SALMAutomodel` — a new SALM variant that uses [NeMo Automodel](https://github.com/NVIDIA-NeMo/Automodel) instead of HuggingFace Transformers for the LLM backbone, enabling efficient training and inference of Speech LLMs with MoE architectures (e.g., Nemotron Nano V3 30B/3B) via native FSDP2, TP, EP, and Grouped GEMM support. CP support will be added later and PP support is not planned.

**Collection**: speechlm2, common

# Changelog

- Add `SALMAutomodel` model class (`nemo/collections/speechlm2/models/salm_automodel.py`) — a `LightningModule` that replaces HF `AutoModelForCausalLM` with NeMo Automodel's distributed LLM loader, supporting deferred initialization via `configure_model()` for memory-efficient per-GPU shard loading.
- Add `AutomodelParallelStrategy` (`nemo/collections/speechlm2/parts/parallel.py`) — a custom Lightning `ModelParallelStrategy` that delegates device mesh creation to NeMo Automodel, supporting FSDP2, TP, PP, CP, EP (MoE), and HSDP.
- Add native Automodel LoRA support (`nemo/collections/speechlm2/parts/automodel_lora.py`) — replaces HuggingFace PEFT with Automodel's built-in `ModuleMatcher`-based LoRA, applied before FSDP2 sharding for correct meta-device handling.
- Add `FlashPrecision` plugin (`nemo/utils/trainer_utils.py`) — a new `bf16-flash`/`fp16-flash` precision mode that sets module dtype without mutating `torch.set_default_dtype`, preventing dtype conflicts with Automodel's internal initialization.
- Add `FlashOptim` compatibility patch (`nemo/core/optim/flash_optim.py`) — fixes `DTensor.from_local()` shape inference for unevenly-sharded FSDP2 parameters during DCP save/load.
- Add `NemotronNanoV3PromptFormatter` (`nemo/collections/common/prompts/nemotron_nano_v3.py`) with `<think>` reasoning support.
- Add dataloader DP rank resolution patch for Automodel's device mesh (handles `dp_replicate`/`dp_shard` naming).
- Add `init_from_checkpoint` support for fine-tuning from DCP directories, HuggingFace directories, and single-file `.ckpt` checkpoints.
- Add `to_hf.py` support for distributed models (gathering DTensor shards before export).
- Add `speechlm2` pip extra in `pyproject.toml`/`setup.py` and uv index configuration for torch CUDA sources.
- Add SALMAutomodel tutorial notebook (`tutorials/speechlm2/SpeechLM_With_NeMo_Automodel.ipynb`).
- Add SALMAutomodel documentation in `docs/source/speechlm2/` (models, configs, training/scaling, mixed precision).
- Add example config `examples/speechlm2/conf/salm_automodel.yaml` with Nemotron Nano V3 defaults.
- Add extensive test suite: `test_salm_automodel.py`, `test_salm_automodel_lora.py`, `test_parallel.py`, `test_datamodule_parallel.py`, `test_init_from_checkpoint.py`, `test_flash_optim.py`, `test_flash_precision.py`, `test_nemotron_nano_v3_prompt_formatter.py`, and CI integration test script.
- Separate `SALM` and `SALMAutomodel` into independent classes (no inheritance), selectable via `model.use_nemo_automodel` config flag.

# Usage

```python
# Training with Automodel backend (FSDP2 + Expert Parallelism for MoE)
python examples/speechlm2/salm_train.py \
    --config-path conf \
    --config-name salm_automodel \
    ++model.use_nemo_automodel=true \
    ++model.pretrained_llm=nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-BF16 \
    trainer.devices=8

# With LoRA
python examples/speechlm2/salm_train.py \
    --config-path conf \
    --config-name salm_automodel \
    ++model.lora.dim=128 \
    ++model.lora.alpha=256 \
    ++model.lora.target_modules='["q_proj","v_proj"]'

# Inference — Step 1: Convert checkpoint to HuggingFace format
# Single-file checkpoint:
python examples/speechlm2/to_hf.py \
    class_path=nemo.collections.speechlm2.models.SALMAutomodel \
    ckpt_path=/path/to/checkpoint.ckpt \
    ckpt_config=/path/to/hparams.yaml \
    output_dir=/path/to/hf_checkpoint

# Distributed checkpoint (use same GPU count as training):
torchrun --nproc-per-node=8 examples/speechlm2/to_hf.py \
    class_path=nemo.collections.speechlm2.models.SALMAutomodel \
    ckpt_path=/path/to/distributed_ckpt_dir \
    ckpt_config=/path/to/hparams.yaml \
    output_dir=/path/to/hf_checkpoint

# Inference — Step 2: Run generation from the converted HF checkpoint
python examples/speechlm2/salm_generate.py \
    pretrained_name=/path/to/hf_checkpoint \
    inputs=/path/to/manifest.jsonl

# Distributed inference with model parallelism:
torchrun --nproc-per-node=8 examples/speechlm2/salm_generate.py \
    pretrained_name=/path/to/hf_checkpoint \
    inputs=/path/to/manifest.jsonl \
    tp_size=1 ep_size=8

```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- `nemo_automodel` is added as an optional dependency via the `speechlm2` pip extra (`pip install nemo_toolkit[speechlm2]`).
- The `FlashPrecision` plugin (renamed from `AutomodelPrecision`) is placed in `nemo/utils/trainer_utils.py` for reuse by other collections (e.g., TTS) in subsequent PRs.
- `uv.lock` changes are large due to adding the `nemo_automodel` dependency tree; the actual code changes are ~5,966 lines added across 59 files (excluding lock file).

### CODE REVIEW

**Comment by pzelasko** [comment]
Trying to decide if we should make SALM backward compatible with vanilla `transformers` LLMs (shares lot of logic but gets somewhat complex) or copy this into a new class (cleaner but more duplication). In any case `canary-qwen-2.5b` released checkpoint must work with the final shape of this PR.

**Comment by desh2608** [comment]
> Trying to decide if we should make SALM backward compatible with vanilla `transformers` LLMs (shares lot of logic but gets somewhat complex) or copy this into a new class (cleaner but more duplication). In any case `canary-qwen-2.5b` released checkpoint must work with the final shape of this PR.

(copying my comment from Slack here) In the current PR, does it already work with HF Automodel and NeMo Automodel? If yes, it looks fine to me. Most of the complexity around model loading seems to be in configure_model() and some utility functions in pretrained.py . Other than that, the annoying thing is to have to put DTensor to a full tensor for some operations (I had to do the same for adding audio generation head), but I think it's not too bad.

File: nemo/collections/speechlm2/parts/parallel.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        dp_size=dp_size,
+        dp_replicate_size=dp_replicate_size,
+        distributed_config=distributed_config,
+        moe_config=moe_config,
+    )
+    strategy.create_device_mesh()
+    return strategy
+
+
+class AutomodelParallelStrategy(ModelParallelStrategy):
```
Comment (nithinraok) [inline nemo/collections/speechlm2/parts/parallel.py:67]: Shouldn;t these be at common level? so other collections could use?

File: docs/source/speechlm2/intro.rst [hunk 2]
```diff
@@ -310,22 +351,35 @@ Alternatively, you can train a model using the provided training scripts in the
...
       batch_size=64 \
       max_new_tokens=128 \
       output_manifest=generations.jsonl
 
+To train the SALMAutomodel variant (with NeMo Automodel backend), use the ``salm_automodel`` config:
+
+.. code-block:: bash
+
+    # Train SALMAutomodel with NVIDIA Nemotron Nano V3 MoE backbone on 8 GPUs
+    torchrun --nproc_per_node=8 examples/speechlm2/salm_train.py \
```
Comment (nithinraok) [inline docs/source/speechlm2/intro.rst:367]: can we add another example for launching on multi-node as well

File: examples/speechlm2/conf/salm_automodel.yaml [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    foreach: true # set to false if having issues with tensor-parallelism
+
+  lr_scheduler:
+    _target_: nemo.core.optim.lr_scheduler.CosineAnnealing
+    warmup_steps: 1000
+    min_lr: 1e-6
+    max_steps: ${trainer.max_steps}
+
+trainer:
+  devices: -1
```
Comment (nithinraok) [inline examples/speechlm2/conf/salm_automodel.yaml]: shouldn't we remove `devices` and `num_nodes` from config?

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+                moe_config = MoEParallelizerConfig()
+            automodel_kwargs["distributed_config"] = distributed_config
+            automodel_kwargs["moe_config"] = moe_config
+        if moe_mesh is not None:
+            automodel_kwargs["moe_mesh"] = moe_mesh
+
+        # When LoRA is configured and we have a device_mesh, pass peft_config
+        # through automodel so LoRA is applied before FSDP2 sharding (handles
+        # meta-device init correctly).
+        peft_config = make_peft_config(self.cfg.lora) if "lora" in self.cfg else None
```
Comment (nithinraok) [inline nemo/collections/speechlm2/models/salm_automodel.py:650]: if user passes `lora: null` in config, this will crash I think

File: nemo/collections/common/prompts/qwen.py [hunk 5]
```diff
@@ -36,3 +39,183 @@ class QwenPromptFormatter(PromptFormatter):
             },
         },
     }
+
+
+class Qwen3PromptFormatter(PromptFormatter):
```
[the remark this answers is not in the document]
Comment (stevehuang52) [reply nemo/collections/common/prompts/qwen.py:44]: @pzelasko is claude correct about `PromptFormatter.resolve()`?

File: nemo/collections/speechlm2/models/salm.py [hunk 6]
```diff
@@ -57,9 +57,15 @@ def __init__(self, cfg) -> None:
         self.cfg = DictConfig(cfg)
         self.audio_locator_tag = self.cfg.audio_locator_tag
 
-        self.tokenizer = AutoTokenizer(self.cfg.pretrained_llm, use_fast=True)
+        self.tokenizer = AutoTokenizer(
+            self.cfg.pretrained_llm, use_fast=True, trust_remote_code=self.cfg.get("trust_remote_code", False)
+        )
         self.tokenizer.add_special_tokens({"additional_special_tokens": [self.audio_locator_tag]})
-        self.llm = load_pretrained_hf(self.cfg.pretrained_llm, pretrained_weights=self.cfg.pretrained_weights)
+        self.llm = load_pretrained_hf(
```
Comment (stevehuang52) [inline nemo/collections/speechlm2/models/salm.py:69]: In future PR, let's rename it to `load_pretrained_hf_llm` as we also have speech models on HF...

File: nemo/collections/speechlm2/parts/pretrained.py [hunk 7]
```diff
@@ -65,6 +65,51 @@ def load_pretrained_hf(
         return AutoModelForCausalLM.from_config(config, torch_dtype=dtype, trust_remote_code=trust_remote_code)
 
 
+def load_pretrained_automodel(
```
Comment (stevehuang52) [inline nemo/collections/speechlm2/parts/pretrained.py]: How about `load_pretrained_automodel_llm`?

File: nemo/collections/speechlm2/parts/parallel.py (hunk 1 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/parts/parallel.py:67]: Good feedback, refactored it to `FlashPrecision` and set up for future addition to other collections (will do collection-specific changes in a subsequent PR)

File: docs/source/speechlm2/intro.rst (hunk 2 above)
Comment (pzelasko) [reply docs/source/speechlm2/intro.rst:367]: No, this is too infrastructure specific, at least for the scope of this PR. Generally you launch it just the same as any other multi-node training.

File: examples/speechlm2/conf/salm_automodel.yaml (hunk 3 above)
Comment (pzelasko) [reply examples/speechlm2/conf/salm_automodel.yaml]: Why? Every other config has them.

File: nemo/collections/speechlm2/models/salm_automodel.py (hunk 4 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/models/salm_automodel.py:650]: Thanks, fixing.

File: nemo/collections/speechlm2/models/salm.py (hunk 6 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/models/salm.py:69]: OK

File: nemo/collections/speechlm2/parts/pretrained.py (hunk 7 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/parts/pretrained.py]: done

**Review by nithinraok — DISMISSED** [review]
LGTM. 

Pls move common AutoModel classes to common collection for re-use later.
